Project Proposal: Browser Phase 1

tedsuo opened the pull request:
OpenTelemetry is rebooting our telemetry efforts for the browser as a sequence of small, highly focused projects. The goal of this first project will be to create instrumentation for the browser runtime and a small initial set of browser libraries, plus the prerequisite API and data modeling work needed to create this instrumentation.

Currently, the proposed meeting time is Thursdays at 8am Pacific. Comment on the doc as to whether this time works for you.

Still needed:

- [x] SIG Sponsor
- [x] Approval for meeting time
- [x] Project Board

svrnm commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,155 @@
+
+* Loading and unloading in the browser are specialized tasks that are very different from booting and shutting down a NodeJS application. Unloading in particular is under strict time and resource constraints.
+* Memory, cpu, and networking resources are severely constrained, especially on mobile devices.
+* A lack of compression and gRPC in browsers
```
```suggestion
* A lack of compression and gRPC in browsers
* Sending and offloading telemetry data requires additional considerations, since browser windows can be closed at any time
```

Maybe worth to call this out additionally as well? Or is [beacon](https://caniuse.com/beacon) good enough these days so there is nothing to worry about anymore?

tedsuo replied:
This is listed above under "Loading and unloading"

svrnm replied:
ah, then I missed that, thanks!

martinkuba-sn commented:
8am Pacific on Thursdays is in general fine with me 👍

martinkuba commented:
8am Pacific on Thursdays is in general fine with me 👍

tedsuo commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,155 @@
+
+Rather than start from scratch with a new SDK, we would like to focus our browser work on instrumentation.  OTel instrumentation has only one dependency, the OTel instrumentation API. There is no need to do SDK work in order to stabilize the instrumentation packages we want to provide for our community.
+
+We propose that the various existing browser clients bind to the new OTel Browser API, and we delay work on providing an optimized SDK solution until after we have stabilized a key set of instrumentation packages. This will allow the community to begin receiving value from our OTel browser work much faster than if we were to start by optimizing the SDK.
```
One core issue to investigate is whether the current API packages work well in the browser, and what can be done to reduce their package sizes. There's also places where the browser is missing features which are present in NodeJS.

I will clarify that the goal is not to cause the APIs to diverge from the spec, but to make them load better in the browser.

tedsuo commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,155 @@
+### Support
+
+We will also decide our compatibility story.
+* Which versions of which browsers do we plan to officially support?
```
Thanks David. If a feature based approach is better we can explore that. I think it would still be good to publish the resulting version matrix.

tedsuo commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,155 @@
+
+During this phase, the SIG will not be focused on implementing a browser-optimized SDK. Instead, we will continue to use the existing OTel JS SDK as our reference implementation. Third party SDKs may also choose to bind to the OTel Browser API at this time.
+
+Instrumentation for the browser runtime:
```
That is a good question. I think it will need to be on a case by case basis, depending on whether the new instrumentation will be unstable or need to be packaged differently.

tedsuo commented:
Thanks for the feedback everyone, I've updated the proposal in response.

* Clarified intentions around the API packages
* Clarified our goals relating to code management and existing instrumentation
* Added more members.

I also updated the meeting time to 8:30, with a 30 min meeting being the goal. We are going to experiment with shorter meetings to see if that allows us to move faster and with clearer intentions.

I am OOO until the 27th, I will respond to further feedback at that time. If we have enough approvals, I would like to start the SIG the following week, with June 5th as the kick off meeting.

danielgblanco commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,161 @@
+* Dependencies
+* Other browser requirements that differ from NodeJS
+
+Data Modelling:
```
Would this work need cross-SIG alignment with @open-telemetry/semconv-client-approvers? Or is this more focused towards the implementation of common semconv?

We are trying to decide if the current Client-Side SIG needs to continue existing as a SIG, or if those conversations about cross-SIG semconv can be folded onto the general semconv meeting + Slack channel for async conversations.

For now, we've decided to keep the Client-Side SIG meetings and move it to a fortnightly meeting after the Browser SIG starts.

tedsuo replied:
There are definitely some cross-platform client issues still open. Sessions and user interactions come to mind. Maybe these discussions will pick up speed once the browser SIG starts, but for now it feels like shutting down the Client SIG (or moving to a monthly meeting) is the right choice, as there does not appear to be much active discussion going on.

jpkrohling reviewed (changes requested):
I'm supportive of this but I think it should have the same requirements as any other new project. I'm specifically looking for this, from the project template:

> Industry outreach
> Who (people, companies) in the industry should be aware of this effort? Was there an attempt to get them onboard? What did they say?

vigneshshanmugam commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,161 @@
+	
+**NodeJS Maintainers and Approvers:**
+
+When bootstrapping our browser work, we want to make sure that the existing OpenTelemetry Javascript community is involved. Existing OTel JS maintainers and approvers should participate in this project to help ensure that it is successful.
```
Whats the path to add more folks involved? Is it similar to the existing process?

tedsuo replied:
Is there specific work you are interested in taking on?

tedsuo replied:
@vigneshshanmugam Thank you! Would you be willing to be listed as an implementation engineer? Meaning that you will make some time to write code in addition to reviewing designs?

vigneshshanmugam replied:
Missed this comment, yes happy to get involved in both the implementation and design. Thanks for the poke.

vigneshshanmugam commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,161 @@
+
+* Loading and unloading in the browser are specialized tasks that are very different from booting and shutting down a NodeJS application. Unloading in particular is under strict time and resource constraints.
+* Memory, cpu, and networking resources are severely constrained, especially on mobile devices.
+* A lack of compression and gRPC in browsers
```
Just a note - Compression spec is already available in most of the browsers https://developer.mozilla.org/en-US/docs/Web/API/Compression_Streams_API, also leveraged by existing RUM tools.

tedsuo replied:
Thanks, good to know!

vigneshshanmugam commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,161 @@
+Options include:
+* Continue implementing instrumentation packages.
+* Implement a browser-specific SDK.
+* Implement a more efficient client protocol.
```
Not sure if we need it to explicit, just for clarity - what does this mean in practise? There are existing techniques that are used in the browsers that works by shortening the field names, but now with the compression stream spec - the need for that is reduced. These techniques were done without shipping a GRPC specific package to the client as that would influence the package size shipped to users.

danielgblanco replied:
We're talking about OTLP here. For example, currently, OTLP requires different exporters and connections per signals (i.e. one for traces, one for logs, etc). These connections would be coalesced if they go to the same origin, but regardless, certain aspects like _Resource_ attributes need to be duplicated in batches exported for each of those signals, whereas if we had multi-signal OTLP we could do it in one batch. This is just one example, but I expect there will be more things that can be further optimised.

vigneshshanmugam replied:
Thanks for the extra context, makes sense 👍🏽

tedsuo commented:
> I'm supportive of this but I think it should have the same requirements as any other new project. I'm specifically looking for this, from the project template:
> 
> > Industry outreach
> > Who (people, companies) in the industry should be aware of this effort? Was there an attempt to get them onboard? What did they say?

Thanks @jpkrohling.  I updated the proposal to include this section.

tedsuo commented:
> I think this is long overdue and would be happy to sponsor. As a JS maintainer I think I am uniquely qualified to help evaluate what can be changed in the existing JS implementation vs what would need to be reimplemented.

Great! I added you as a sponsor. Given that much of this SIG is specifically focused on thorny issues related to differences between the NodeJS and Browser runtime, I agree that a sponsor such as yourself is critical to the success of this project.

I've also added @carlosalberto as an escalating sponsor to fulfill the TC requirement.

jpkrohling commented on projects/browser-phase-1.md:
```diff
@@ -90,15 +90,20 @@ Options include:
 
+### Industry Outreach
+
+There are a number of existing RUM/Client observability implementations. Representatives from a wide selection of companies that have experience with browser observability are present in the SIG. This includes Microsoft, New Relic, DynaTrace, HoneyComb, Grafana Labs, and Cisco, among others. We plan for further outreach among end users once we have working code for them to review.
```
Some clear leaders in the area are missing. My concern is that we'll, once again, design something without hearing opinions from people we **know** to be relevant in the market.

I realize that some of those voices, like Sentry, have not been exactly friendly towards OTel in the past, which is yet another reason to reach out to them right now: they clearly have opinions on how improve what we have right now in the browser use-case.

I know nothing about this world of RUM, but a quick search also shows me that some of the apparent leaders in the area aren't in the list above. Companies like Datadog, Sematext, PostHog, Akamai, ... I'm confident you know who's relevant in the market, given how close you are to the SIG and this area, so I'd be interested in hearing what they (and others) said when asked to join this effort.

jpkrohling commented on projects/browser-phase-1.md:
```diff
@@ -90,15 +90,20 @@ Options include:
+* Daniel Dyla - JS Maintainer Sponsor
+* Carlos Alberto Cortez – TC Escalating Sponsor
 
 **Implementation Engineers:**
```
I understand that Martin is from ServiceNow and Purvi from Honeycomb. Didn't the other companies in your list commit to help in the implementation? I see Marco (Grafana) and David (Elastic) in the list of NodeJS maintainers and approvers here (apparently, it doesn't match with the list I see on opentelemetry-js?): are they also committing to work on the implementation?

tedsuo replied:
Ok, I can review the implementation engineers section with the client SIG and update it with more commitments.

I'm not sure what you mean about NodeJS approvers section? Everyone listed in that section is part of the javascirpt approvers team. Can you clarify?

jpkrohling reviewed (changes requested):
Sorry for not approving this yet, I want to make sure the chances of success for this project are higher this time.

dyladan commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,166 @@
+
+**Sponsors:**
+* Ted Young - Spec Maintainer Sponsor
+* Daniel Dyla - JS Maintainer Sponsor
```
I don't think implementation SIG maintainers are necessarily allowed to sponsor other SIGs. I'm offering in my capacity as spec sponsor. I just think being the JS maintainer makes me particularly relevant for this role.

```suggestion
* Daniel Dyla - Spec Sponsor
```

tedsuo replied:
Gotcha!

reyang commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,166 @@
+
+## Goals, Objectives, and Requirements
+
+OpenTelemetry needs a javascript implementation optimized for the browser. Sharing code with NodeJS is helpful when possible. But going forward, we want to prioritize running well in the browser over code reuse with the NodeJS project. If there are places where code reuse with NodeJS is a hard requirement, those places should be identified explicitly.
```
This is vague, we should be explicit about the technical direction of API package - which approach will this SIG take?

* The SIG aims to produce a shared OpenTelemetry API package that can be used by both web browsers and Node.js services.
* The SIG aims to produce two separate OpenTelemetry API packages, one for web browser and one for Node.js services.
* The SIG doesn't know the answer yet and would expect to figure this out later.

There are many libraries (e.g., NPM packages) that target both web browsers and Node.js services, these libraries will need an OpenTelemetry API package for instrumentation.

dyladan replied:
My personal opinion is that a separate web and node API is a non-starter for exactly this reason:

> There are many libraries (e.g., NPM packages) that target both web browsers and Node.js services, these libraries will need an OpenTelemetry API package for instrumentation.

I would much prefer to improve the existing API package even if that means a 2.0

dyladan replied:
we can discuss this in the SIG though

danielgblanco commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,169 @@
+**Design review:**
+* Ram Thiru (Microsoft)
+* Santosh Cheler (Cisco / Splunk)
+* Marco Schäfer (Grafana Labs)
```
```suggestion
* Marco Schäfer (Grafana Labs)
* Jordan Porter (New Relic)
```
cc @metal-messiah

reyang commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,169 @@
+
+## Current Challenges
+
+OpenTelemetry currently has a NodeJS-focused javascript implementation that is also capable of running in the browser. However, the requirements for a successful browser observability solution differ enough from NodeJS that a specialized solution is required.
```
I don't see Web Assembly being mentioned, maybe that'll be Phase 3.

I want to call out that the browser observability story isn't just about JavaScript. In other words, if this group designed some new APIs just for JavaScript in this Browser Phase 1, we might soon realize that we need to redesign these APIs for web assemblies.

tedsuo replied:
Correct, we are not working on web assembly during this phase of development. I disagree that web assembly would have any bearing on the OpenTelemetry APIs we are designing here for javascript applications running in the browser. Can you explain why it would be relevant?

reyang replied:
It is probably a question regarding "is web assembly part of the Browser project or not?". I can see three answers:
1. It is. Web assembly will be considered in a later phase, not Browser Phase 1 though.
2. It is not. Web assembly is a totally separate topic, and we are not going to consider it as part of the Browser support (although web assembly works for web browsers).
3. We don't know the answer yet, we just want to not think/worry about it for now.

tedsuo replied:
Ok, thank you for the clarification. I believe that web assembly would be an entirely separate topic. The issue there is that instrumented code compiled with web assembly would use the OpenTelemetry APIs in that language. For example, a C++ library would be instrumented with Otel C++, not with Otel Browser JS.

On its own, a complete web assembly application using OpenTelemetry would probably work. However, it is not clear to me how context propagation would work between separate libraries compiled using web assembly and then mixed together. Not to mention, sharing an SDK. There are probably a lot of issues with this approach. Most likely, this would be it's own specialized implementation.

I will list it as a potential future topic, and as something to consider when we review the API design. But I don't think that we should focus on it as an implementation target at this time. In general, we are trying to keep this project tightly scoped. For example, even an SDK for the browser is out of scope.

carlosalberto replied:
Is this resolved @reyang ?

reyang commented on projects/browser-phase-1.md:
```diff
@@ -0,0 +1,171 @@
+* @ramthi (Microsoft)
+* @scheler (Cisco / Splunk)
+* @codecapitano (Grafana Labs)
+* metal-messiah (New Relic)
```
```suggestion
* @metal-messiah (New Relic)
```
